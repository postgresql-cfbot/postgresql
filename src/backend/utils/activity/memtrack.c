/*-------------------------------------------------------------------------
 *
 * memtrack.c
 *	  track and manage memory usage by the PostgreSQL server.
 *
 * Portions Copyright (c) 1996-2023, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  src/backend/utils/activity/memtrack.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"
#include "miscadmin.h"
#include "utils/backend_status.h"
#include "utils/memtrack.h"
#include "storage/proc.h"
#include "storage/pg_shmem.h"
#include "utils/pgstat_internal.h"

/*
 * Max backend memory allocation allowed (MB). 0 = disabled.
 * Max backend bytes is the same but in bytes.
 * These default to "0", meaning don't check bounds for total memory.
 */
int			max_total_memory_mb = 0;
int64		max_total_memory_bytes = 0;

/*
 * Private variables for tracking memory use.
 * These values are preset so memory tracking is active on startup.
 * After a fork(), they must be reset using 'fork_tracked_memory()'.
 */
PgStat_Memory my_memory = INITIAL_ALLOCATED_MEMORY;
PgStat_Memory reported_memory = NO_ALLOCATED_MEMORY;
int64		reservation_lower_bound = 0;
int64		reservation_upper_bound = 0;

/*
 * Reset memory tracking after a fork.
 * We actually keep the memory intact, but
 * the memory hasn't been added to the global totals.
 *
 * The counters are properly initialized at startup,
 * so this function only needs to be called after a fork().
 */
void
fork_tracked_memory(void)
{
	/* This new process hasn't reported any memory yet. */
	reported_memory = NO_ALLOCATED_MEMORY;

	/* Force allocations to be reported once ProcGlobal is initialized. */
	reservation_lower_bound = 0;
	reservation_upper_bound = 0;

	/* Release the DSM reservations since we didn't create them. */
	update_local_reservation(-my_memory.subTotal[PG_ALLOC_DSM], PG_ALLOC_DSM);
}

/*
 * Clean up memory counters as backend is exiting.
 *
 * DSM memory is not automatically returned, so it persists in the counters.
 * All other memory will disappear, so those counters are set to zero.
 *
 * Ideally, this function would be called last, but in practice there are some
 * late memory releases that happen after it is called.
 */
void
exit_tracked_memory(void)
{
	/*
	 * Release all of our private (non-dsm) memory. We don't release dsm
	 * shared memory since it survives process exit.
	 */
	for (int type = 0; type < PG_ALLOC_TYPE_MAX; type++)
		if (type != PG_ALLOC_DSM)
			update_local_reservation(-my_memory.subTotal[type], type);

	/* Report the final values to shmem (just once) */
	(void) update_global_reservation(0, 0);

	/*
	 * Sometimes we get late memory releases after this function is called.
	 * We've already reported all our private memory as released. Set the
	 * bounds to ensure we don't report those late releases twice.
	 */
	reservation_lower_bound = INT64_MIN;
	reservation_upper_bound = INT64_MAX;
}


/*
 * Update memory reservation for a new request.
 *
 * There are two versions of this function. This one, which updates
 * global values in shared memory, and an optimized update_local_reservation()
 * which only updates private values.
 *
 * This routine is the "slow path". We invoke it periodically to update
 * global values and pgstat statistics.
 *
 * We also invoke it whenever we reserve DSM memory. This ensures the
 * DSM memory counter is up-to-date, and more important, ensures it
 * never goes negative.
 */
bool
update_global_reservation(int64 size, pg_allocator_type type)
{
	int64		delta;
	uint64		dummy;
	PgStatShared_Memtrack *global = &pgStatLocal.shmem->memtrack;

	/*
	 * If we are still initializing, only update the private counters.
	 * The tests are:
	 *   1) Is pg shared memory attached?
	 *   2) Are statistics initialized?
	 *   3) Is postmaster up and running?
	 *   4) If backend, is MyBEEntry set up?
	 */
	if (UsedShmemSegAddr == NULL || pgStatLocal.shmem == NULL || PostmasterPid == 0 ||
		(MyProcPid != PostmasterPid && MyBEEntry == NULL))
		return update_local_reservation(size, type);

	/* Verify totals are not negative. This is both a pre- and post-condition. */
	Assert((int64) pg_atomic_read_u64(&global->total_memory_reserved) >= 0);
	Assert((int64) pg_atomic_read_u64(&global->total_dsm_reserved) >= 0);

	/* Calculate total bytes allocated or freed since last report */
	delta = my_memory.total + size - reported_memory.total;

	/*
	 * If memory limits are set, we are increasing our reservation and we
	 * are not the postmaster...
	 */
	if (max_total_memory_bytes > 0 && size > 0 && MyProcPid != PostmasterPid && delta > 0)
	{
		/* Update the global total memory counter subject to the upper limit. */
		if (!pg_atomic_fetch_add_limit_u64(&global->total_memory_reserved, delta, max_total_memory_bytes, &dummy))
			return false;
	}

	/*
	 * Otherwise, update the global counter with no limit checking.
	 */
	else
		(void) pg_atomic_fetch_add_u64(&global->total_memory_reserved, delta);

	/*
	 * Update the private memory counters. This must happen after the limit is
	 * checked.
	 */
	(void) update_local_reservation(size, type);

	/*
	 * Update the global dsm memory counter. Since we always take this path
	 * when dsm memory is allocated, the reported value is up-to-date, and we
	 * can simply add in the new size. We don't need to calculate the delta as
	 * we do for private memory allocators.
	 */
	if (type == PG_ALLOC_DSM)
		(void) pg_atomic_fetch_add_u64(&global->total_dsm_reserved, size);

	/* Report the current memory allocations for either postmaster or backend */
	if (MyProcPid == PostmasterPid)
		pgstat_report_postmaster_memory();
	else
		pgstat_report_backend_memory();

	/* Remember the values we just reported */
	reported_memory = my_memory;

	/* Update bounds so they bracket our new allocation size. */
	reservation_upper_bound = my_memory.total + allocation_allowance_refill_qty;
	reservation_lower_bound = my_memory.total - allocation_allowance_refill_qty;

	/*
	 * Verify totals are not negative. By checking as a post-condition, we are
	 * more likely to identify the code that caused the problem.
	 */
	Assert((int64) pg_atomic_read_u64(&global->total_memory_reserved) >= 0);
	Assert((int64) pg_atomic_read_u64(&global->total_dsm_reserved) >= 0);

	return true;
}
