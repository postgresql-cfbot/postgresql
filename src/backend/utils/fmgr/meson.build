backend_sources += files(
  'dfmgr.c',
  'fmgr.c',
  'funcapi.c',
)

# fmgrtab.c
generated_backend_sources += fmgrtab_target[2]
