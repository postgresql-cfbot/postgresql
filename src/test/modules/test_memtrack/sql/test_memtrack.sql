
-- verify the pg_stat_memory_allocation view exists
SELECT
        pid > 0, total_reserved >= 0, init_reserved >= 0, aset_reserved >= 0, dsm_reserved >= 0, generation_reserved >= 0, slab_reserved >= 0
FROM
    pg_stat_memory_reservation limit 1;

-- verify the pg_stat_global_memory_tracking view exists
SELECT
        total_memory_reserved >= 0, dsm_memory_reserved >= 0, total_memory_available >= 0, static_shared_memory >= 0
FROM
    pg_stat_global_memory_tracking;

-- verify some common backends have reserved memory
SELECT
        total_reserved >= 0 AS result
FROM
    pg_stat_activity ps
        JOIN pg_stat_memory_reservation pa ON (pa.pid = ps.pid)
WHERE
        backend_type IN ('checkpointer', 'background writer', 'walwriter', 'autovacuum launcher');


-- For each process, the total should be the sum of subtotals
SELECT *
FROM
    pg_stat_memory_reservation
WHERE total_reserved != (init_reserved + aset_reserved + dsm_reserved + generation_reserved + slab_reserved);

-- For each process, the initial allocation is >= 1 MB
SELECT *
FROM
    pg_stat_memory_reservation
WHERE
    init_reserved < 1024*1024;

-- For current backend, the top context should match the sum from the allocators.
SELECT *
FROM
    pg_backend_memory_allocation
WHERE total_top_context_allocated <> (aset_allocated + generation_allocated + slab_allocated);

CREATE EXTENSION test_memtrack;

-- Make sure we can track memory usage of a single task
-- Since logic is the same for all memory managers, only test one.
SELECT test_memtrack(1, 1, 0, 1024);

-- Make sure we can track memory usage of multiple tasks.
--  By default we are limited to 8 tasks, so stay below the limit.
SELECT test_memtrack(5, 3, 1024, 5*1024);

-- Do it again. We had a bug where the second call would fail.
SELECT test_memtrack(5, 3, 1024, 5*1024);

-- Now we're going to actually do memory allocations.
-- We'll test each type of memory allocator.

-- Verify we can create and destroy contexts.
SELECT test_allocation(1,1,0,1024);
SELECT test_allocation(1,2,0, 1024);
SELECT test_allocation(1,3, 0,1024);
SELECT test_allocation(1,4,0,1024);

-- Create and free blocks of memory.
SELECT test_allocation(5,1,5*1024,1024);
SELECT test_allocation(5,2,5,1024*1024);  /* Fewer, don't exceed shmem limit */
SELECT test_allocation(5,3,5*1024,1024);
SELECT test_allocation(5,4,5*1024,1024);

-- Add up the private memory for each process and compare with the global total
--  The delta should be 0.
SELECT ABS(process_private - global_private) as delta
FROM
    (SELECT SUM(total_reserved - dsm_reserved)                             AS process_private
    FROM pg_stat_memory_reservation as p),
    (SELECT (total_memory_reserved - dsm_memory_reserved - static_shared_memory) AS global_private
    FROM pg_stat_global_memory_tracking as g);

-- Verify the global dsm memory is at least the sum of processes dsm memory.
-- The global can be larger if some process pinned dsm and than exited.
SELECT *
FROM
    (SELECT SUM(dsm_reserved) as process_dsm from pg_stat_memory_reservation),
    (SELECT dsm_memory_reserved as global_dsm from pg_stat_global_memory_tracking)
WHERE
    global_dsm < process_dsm;

-- Allocate more memory than we have available.
-- (this should fail because we configured max_total_memory to 1024 Mb)
SELECT test_memtrack(5, 2, 1024, 1024*1024*1024);
