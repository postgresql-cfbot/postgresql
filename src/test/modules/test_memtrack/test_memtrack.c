/*--------------------------------------------------------------------------
 *
 * test_memstack.c
 *	Testing the memory tracking features.
 *	Creates a pool of workers which allocate/free memory,
 *	and verifies the totals are as expected.
 *
 * Copyright (c) 2013-2023, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		src/test/modules/test_memtrack/test_memtrack.c
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"

#include "fmgr.h"
#include "miscadmin.h"
#include "pgstat.h"
#include "varatt.h"

#include "storage/dsm.h"
#include "utils/backend_status.h"
#include "utils/memutils_internal.h"
#include "utils/memtrack.h"

#include "worker_pool.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(test_memtrack);
PG_FUNCTION_INFO_V1(test_allocation);

/* Message sent to the worker */
typedef enum
{
	ALLOCATE, RELEASE
}			Action;
typedef struct AllocateRequest
{
	Action		action;
	pg_allocator_type type;
	int32		nWorkers;
	int32		nBlocks;
	int64		blockSize;
}			AllocateRequest;

/* Message received back from the worker */
typedef struct ResponseData
{
	int32		errorCode;
	PgStat_Memory memory;
	PgStat_Memory startingMemory;
}			ResponseData;

/* Forward references */
static bool reserveBlocks(pg_allocator_type type, int nBlocks, int blockSize);
static bool releaseBlocks(pg_allocator_type type, int nBlocks, int blockSize);
static void validateArgs(int nWorkers, pg_allocator_type type, int nBlocks, int blockSize);
static void sendRequest(WorkerPool * pool, int worker, Action action, pg_allocator_type type, int nBlocks, int blockSize);
static void processReply(WorkerPool * pool, int worker, Action actions, pg_allocator_type type, int nBlocks, int blockSize);
static Datum exercise_worker(FunctionCallInfo fcinfo, char *workerFunction);
static void checkAllocations();

/* Test the memory tracking features standalone */
PGDLLEXPORT Datum test_memtrack(PG_FUNCTION_ARGS);
PGDLLEXPORT void test_memtrack_worker(Datum arg);

/* Test the memory tracking features with actual allocations */
PGDLLEXPORT Datum test_allocation(PG_FUNCTION_ARGS);
PGDLLEXPORT void test_allocation_worker(Datum arg);

/*
 * Test the memory tracking features.
 */
Datum
test_memtrack(PG_FUNCTION_ARGS)
{
	return exercise_worker(fcinfo, "test_memtrack_worker");
}


Datum
test_allocation(PG_FUNCTION_ARGS)
{
	return exercise_worker(fcinfo, "test_allocation_worker");
}



/*
 * Test the memory tracking features
 * Creates a pool of workers, issues requests, and verifies the results.
 */
static
Datum
exercise_worker(FunctionCallInfo fcinfo, char *workerFunction)
{
	WorkerPool	pool[1];
	int64		delta,
				starting_bkend_bytes;
	int64		expected,
				fudge;
	PgStatShared_Memtrack *global = &pgStatLocal.shmem->memtrack;

	/* Get  the SQL function arguments */
	int32		nWorkers = PG_GETARG_INT32(0);
	pg_allocator_type type = PG_GETARG_INT32(1);
	int32		nBlocks = PG_GETARG_INT32(2);
	int32		blockSize = PG_GETARG_INT32(3);

	validateArgs(nWorkers, type, nBlocks, blockSize);

	/* Our global totals may be off by an allocation allowance per worker */
	fudge = nWorkers * allocation_allowance_refill_qty * nWorkers;
	expected = nWorkers * nBlocks * blockSize;

	/* Set up pool of background workers */
	initWorkerPool(pool, nWorkers, sizeof(AllocateRequest) * 10, sizeof(ResponseData) * 10,
							"test_memtrack", workerFunction);

	/* Remember the total memory before we start allocations */
	starting_bkend_bytes = pg_atomic_read_u32((void *) &global->total_memory_reserved);

	/* Tell the workers to start their first batch of allocations */
	for (int w = 0; w < nWorkers; w++)
		sendRequest(pool, w, ALLOCATE, type, nBlocks, blockSize);

	/* Get the workers response and verify all is good */
	for (int w = 0; w < nWorkers; w++)
		processReply(pool, w, ALLOCATE, type, nBlocks, blockSize);

	/* Confirm the total backend memory is greater than what we just allocated */
	delta = pg_atomic_read_u32((void *) &global->total_memory_reserved) - starting_bkend_bytes;
	if (delta < expected - fudge || delta > expected + fudge)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("Total allocated memory %zd doesn't match expected %zd  fudge=%zd\n", delta, expected, fudge)));

	/* Tell the workers to release their memory */
	for (int w = 0; w < nWorkers; w++)
		sendRequest(pool, w, RELEASE, type, nBlocks, blockSize);

	/* Get the workers response and verify all is good */
	for (int w = 0; w < nWorkers; w++)
		processReply(pool, w, RELEASE, type, nBlocks, blockSize);

	/* Verify the new total is reasonable */
	delta = pg_atomic_read_u32((void *) &global->total_memory_reserved) - starting_bkend_bytes;
	if (delta < -fudge || delta > fudge)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("Total allocated memory %zd is less than expected %d\n", delta, 0)));

	/* Clean up */
	freeWorkerPool(pool);
	PG_RETURN_VOID();
}


/*
 * Verify the arguments passed to the SQL function are valid.
 */
static
void
validateArgs(int nWorkers, pg_allocator_type type, int nBlocks, int blockSize)
{
	/* A negative blockcount is nonsensical. */
	if (nBlocks < 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("repeat count size must be an integer value greater than or equal to zero")));

	/* a minimum of 1 worker is required. */
	if (nWorkers <= 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("number of workers must be an integer value greater than zero")));

	/* block size must be > 0 */
	if (blockSize <= 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("block size must be an integer value greater than zero")));

	/* Valid type? */
	if (type < 0 || type >= PG_ALLOC_TYPE_MAX)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("invalid allocation type")));
}

/*
 * Send a request message to a worker.
 */
static
void
sendRequest(WorkerPool * pool, int worker, Action action, pg_allocator_type type, int nBlocks, int blockSize)
{
	AllocateRequest req;
	int			result;

	req = (AllocateRequest)
	{
		.nBlocks = nBlocks,.action = action,.type = type,.blockSize = blockSize
	};

	result = sendToWorker(pool, worker, &req, sizeof(req));
	if (result != SHM_MQ_SUCCESS)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("could not send message")));
}


/*
 * Receive a reply from a worker and verify the results make sense.
 */
void
processReply(WorkerPool * pool, int worker, Action action, pg_allocator_type type, int nBlocks, int blockSize)
{
	int64		delta;
	ResponseData *rp;
	ResponseData resp[1];
	Size		len;
	int			result;

	/* Receive a message. Returns a pointer to the message and a length */
	result = recvFromWorker(pool, worker, (void *) &rp, &len);
	if (result != SHM_MQ_SUCCESS)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("could not receive message")));

	/* Copy response message in case the buffer isn't aligned */
	memcpy(resp, rp, sizeof(*resp));

	/* Check for outof memory error */
	if (resp->errorCode != 0)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("worker ran out of memory")));

	/* Verify the totals */
	delta = resp->memory.subTotal[type] - resp->startingMemory.subTotal[type];
	if (action == ALLOCATE && delta < nBlocks * blockSize / 2)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("worker(allocate) increased %zd bytes, expected %d", delta, nBlocks * blockSize)));

	if (action == RELEASE && delta > -nBlocks*blockSize / 2)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("worker(release) decreased %zd bytes, expected %d", -delta, -nBlocks * blockSize)));
}


/*
 * Worker which bumps the memtrack counters without actually allocating anything.
 */
void
test_memtrack_worker(Datum arg)
{
	AllocateRequest *req;
	Size		reqSize;

	ResponseData resp[1];
	shm_mq_result result;

	workerInit(arg);

	/*
	 * Now that we're running, make note of how much memory has been already
	 * allocated
	 */

	do
	{

		result = workerRecv((void *) &req, &reqSize);
		if (result != SHM_MQ_SUCCESS)
			break;

		resp->startingMemory = my_memory;

		/* Allocate or release the blocks */
		if (req->action == ALLOCATE && reserveBlocks(req->type, req->nBlocks, req->blockSize))
			resp->errorCode = 0;
		else if (req->action == RELEASE && releaseBlocks(req->type, req->nBlocks, req->blockSize))
			resp->errorCode = 0;
		else
			resp->errorCode = 1;

		/* Get the current memory totals */
		resp->memory = my_memory;

		/* Send the response */
		result = workerSend(resp, sizeof(resp[1]));

	} while (result == SHM_MQ_SUCCESS);

	workerExit(0);
}


bool
reserveBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	/* Allocate the requested number of blocks */
	for (int i = 0; i < nBlocks; i++)
		if (!reserve_tracked_memory(blockSize, type))
			return false;

	return true;
}


bool
releaseBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	for (int i = 0; i < nBlocks; i++)
		release_tracked_memory(blockSize, type);

	return true;
}

/* Forward references for the allocation worker */
static bool freeBlocks(pg_allocator_type type, int nBlocks, int blockSize);
static MemoryContext createTestContext(pg_allocator_type type);
static bool allocateBlocks(pg_allocator_type type, int nBlocks, int blockSize);
static bool allocateDSMBlocks(int nBlocks, int blockSize);
static bool freeDSMBlocks(int nBlocks, int blockSize);
static bool allocateContextBlocks(pg_allocator_type type, int nBlocks, int blockSize);
static bool freeContextBlocks(pg_allocator_type type, int nBlocks, int blockSize);


/* An array of pointers to the allocated blocks */
static void **allocations;
static MemoryContext testContext;


/*
 * Worker which actually allocates and releases memory.
 */
void
test_allocation_worker(Datum arg)
{
	AllocateRequest *req;
	Size		reqSize;

	ResponseData resp[1];
	shm_mq_result result;

	PgStat_Memory startingMemory;

	workerInit(arg);

	do
	{
		result = workerRecv((void *) &req, &reqSize);
		if (result != SHM_MQ_SUCCESS)
			break;

		startingMemory = my_memory;

		/* Allocate or release the blocks */
		if (req->action == ALLOCATE && allocateBlocks(req->type, req->nBlocks, req->blockSize))
			resp->errorCode = 0;
		else if (req->action == RELEASE && freeBlocks(req->type, req->nBlocks, req->blockSize))
			resp->errorCode = 0;
		else
			resp->errorCode = 1;

		/* Get the current memory totals */
		resp->memory = my_memory;
		resp->startingMemory = startingMemory;

		/* Send the response */
		result = workerSend(resp, sizeof(resp[1]));

		checkAllocations();

	} while (result == SHM_MQ_SUCCESS);

	workerExit(0);
}



static bool
allocateBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	if (type == PG_ALLOC_DSM)
		return allocateDSMBlocks(nBlocks, blockSize);
	else
		return allocateContextBlocks(type, nBlocks, blockSize);
}

static bool
freeBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	if (type == PG_ALLOC_DSM)
		return freeDSMBlocks(nBlocks, blockSize);
	else
		return freeContextBlocks(type, nBlocks, blockSize);
}



/*
 * Create a memory context for the non-DSM memory allocations.
 */
static
MemoryContext
createTestContext(pg_allocator_type type)
{
	switch (type)
	{
		case PG_ALLOC_ASET:
			return AllocSetContextCreate(TopMemoryContext, "test_aset", 0, 1024, 16 * 1024);
		case PG_ALLOC_SLAB:
			return SlabContextCreate(TopMemoryContext, "test_slab", 16 * 1024, 1024);
		case PG_ALLOC_GENERATION:
			return GenerationContextCreate(TopMemoryContext, "test_gen", 0, 1024, 16 * 1024);
		default:
			return NULL;

	}
}


/*
 * Allocate blocks of memory from a non-DSM context.
 */
static
bool
allocateContextBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	MemoryContext old;

	if (type == PG_ALLOC_DSM || type == PG_ALLOC_INIT)
		return false;

	/* Create a memory context for the allocations */
	testContext = createTestContext(type);
	if (testContext == NULL)
		return false;

	/* Switch to the new context */
	old = MemoryContextSwitchTo(testContext);

	/* Create a list of block pointers - not in the context */
	allocations = malloc(sizeof(void *) * nBlocks);
	if (allocations == NULL)
		return false;

	/* Allocate the requested number of blocks */
	for (int i = 0; i < nBlocks; i++)
	{
		allocations[i] = palloc(blockSize);
		if (allocations[i] == NULL)
			return false;
	}

	/* Switch back to the old context */
	MemoryContextSwitchTo(old);

	return true;
}


/*
 * Free blocks of memory allocated earlier
 */
static
bool
freeContextBlocks(pg_allocator_type type, int nBlocks, int blockSize)
{
	for (int i = 0; i < nBlocks; i++)
		pfree(allocations[i]);

	MemoryContextDelete(testContext);
	free(allocations);
	allocations = NULL;

	return true;
}


/*
 * Allocate blocks of memory from DSM
 */
static bool
allocateDSMBlocks(int nBlocks, int blockSize)
{

	/* Create a list of block pointers - not in the context */
	allocations = malloc(sizeof(void *) * nBlocks);
	if (allocations == NULL)
		return false;

	for (int i = 0; i < nBlocks; i++)
	{
		allocations[i] = dsm_create(blockSize, 0);
		if (allocations[i] == NULL)
			return false;
	}

	return true;
}


/*
 * Free blocks of DSM memory allocated earlier
 */
static bool
freeDSMBlocks(int nBlocks, int blockSize)
{
	for (int i = 0; i < nBlocks; i++)
		dsm_detach(allocations[i]);

	free(allocations);
	allocations = NULL;


	return true;
}


/*
 * Compare top context allocations with our private memory numbers.
 * They should be the same.
 */
static void
checkAllocations()
{
	int64 my_private = my_memory.total - my_memory.subTotal[PG_ALLOC_INIT] - my_memory.subTotal[PG_ALLOC_DSM];
	int64 context = getContextMemoryTotal();
	Assert(my_private == context);
}
