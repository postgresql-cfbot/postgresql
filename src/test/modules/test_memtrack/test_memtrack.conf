max_total_memory = 8192
