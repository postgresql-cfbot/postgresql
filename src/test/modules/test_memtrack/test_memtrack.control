comment = 'Test code for PostgreSQL memory tracking'
default_version = '1.0'
module_pathname = '$libdir/test_memtrack'
relocatable = true
