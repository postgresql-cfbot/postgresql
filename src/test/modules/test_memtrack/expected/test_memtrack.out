-- verify the pg_stat_memory_allocation view exists
SELECT
        pid > 0, total_reserved >= 0, init_reserved >= 0, aset_reserved >= 0, dsm_reserved >= 0, generation_reserved >= 0, slab_reserved >= 0
FROM
    pg_stat_memory_reservation limit 1;
 ?column? | ?column? | ?column? | ?column? | ?column? | ?column? | ?column? 
----------+----------+----------+----------+----------+----------+----------
 t        | t        | t        | t        | t        | t        | t
(1 row)

-- verify the pg_stat_global_memory_tracking view exists
SELECT
        total_memory_reserved >= 0, dsm_memory_reserved >= 0, total_memory_available >= 0, static_shared_memory >= 0
FROM
    pg_stat_global_memory_tracking;
 ?column? | ?column? | ?column? | ?column? 
----------+----------+----------+----------
 t        | t        | t        | t
(1 row)

-- verify some common backends have reserved memory
SELECT
        total_reserved >= 0 AS result
FROM
    pg_stat_activity ps
        JOIN pg_stat_memory_reservation pa ON (pa.pid = ps.pid)
WHERE
        backend_type IN ('checkpointer', 'background writer', 'walwriter', 'autovacuum launcher');
 result 
--------
 t
 t
 t
 t
(4 rows)

-- For each process, the total should be the sum of subtotals
SELECT *
FROM
    pg_stat_memory_reservation
WHERE total_reserved != (init_reserved + aset_reserved + dsm_reserved + generation_reserved + slab_reserved);
 pid | total_reserved | init_reserved | aset_reserved | dsm_reserved | generation_reserved | slab_reserved 
-----+----------------+---------------+---------------+--------------+---------------------+---------------
(0 rows)

-- For each process, the initial allocation is >= 1 MB
SELECT *
FROM
    pg_stat_memory_reservation
WHERE
    init_reserved < 1024*1024;
 pid | total_reserved | init_reserved | aset_reserved | dsm_reserved | generation_reserved | slab_reserved 
-----+----------------+---------------+---------------+--------------+---------------------+---------------
(0 rows)

-- For current backend, the top context should match the sum from the allocators.
SELECT *
FROM
    pg_backend_memory_allocation
WHERE total_top_context_allocated <> (aset_allocated + generation_allocated + slab_allocated);
 pid | total_top_context_allocated | aset_allocated | dsm_allocated | generation_allocated | slab_allocated 
-----+-----------------------------+----------------+---------------+----------------------+----------------
(0 rows)

CREATE EXTENSION test_memtrack;
-- Make sure we can track memory usage of a single task
-- Since logic is the same for all memory managers, only test one.
SELECT test_memtrack(1, 1, 0, 1024);
 test_memtrack 
---------------
 
(1 row)

-- Make sure we can track memory usage of multiple tasks.
--  By default we are limited to 8 tasks, so stay below the limit.
SELECT test_memtrack(5, 3, 1024, 5*1024);
 test_memtrack 
---------------
 
(1 row)

-- Do it again. We had a bug where the second call would fail.
SELECT test_memtrack(5, 3, 1024, 5*1024);
 test_memtrack 
---------------
 
(1 row)

-- Now we're going to actually do memory allocations.
-- We'll test each type of memory allocator.
-- Verify we can create and destroy contexts.
SELECT test_allocation(1,1,0,1024);
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(1,2,0, 1024);
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(1,3, 0,1024);
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(1,4,0,1024);
 test_allocation 
-----------------
 
(1 row)

-- Create and free blocks of memory.
SELECT test_allocation(5,1,5*1024,1024);
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(5,2,5,1024*1024);  /* Fewer, don't exceed shmem limit */
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(5,3,5*1024,1024);
 test_allocation 
-----------------
 
(1 row)

SELECT test_allocation(5,4,5*1024,1024);
 test_allocation 
-----------------
 
(1 row)

-- Add up the private memory for each process and compare with the global total
--  The delta should be 0.
SELECT ABS(process_private - global_private) as delta
FROM
    (SELECT SUM(total_reserved - dsm_reserved)                             AS process_private
    FROM pg_stat_memory_reservation as p),
    (SELECT (total_memory_reserved - dsm_memory_reserved - static_shared_memory) AS global_private
    FROM pg_stat_global_memory_tracking as g);
 delta 
-------
     0
(1 row)

-- Verify the global dsm memory is at least the sum of processes dsm memory.
-- The global can be larger if some process pinned dsm and than exited.
SELECT *
FROM
    (SELECT SUM(dsm_reserved) as process_dsm from pg_stat_memory_reservation),
    (SELECT dsm_memory_reserved as global_dsm from pg_stat_global_memory_tracking)
WHERE
    global_dsm < process_dsm;
 process_dsm | global_dsm 
-------------+------------
(0 rows)

-- Allocate more memory than we have available.
-- (this should fail because we configured max_total_memory to 1024 Mb)
SELECT test_memtrack(5, 2, 1024, 1024*1024*1024);
ERROR:  worker ran out of memory
